File: irrigation.py
def calculate_irrigation(crop, soil_moisture, temperature, humidity, rainfall_last_7_days, area_hectares=1.0):
    """
    Calculate smart irrigation recommendations using simplified ET₀ approach.
    
    Parameters:
    - crop: crop type string
    - soil_moisture: 0-100 (percent)
    - temperature: current temperature in °C
    - humidity: relative humidity %
    - rainfall_last_7_days: mm in past 7 days
    - area_hectares: farm area
    """
    
    # Crop water requirements (mm/day) by crop
    CROP_ETC = {
        'rice': 8.0, 'wheat': 4.5, 'maize': 5.5, 'cotton': 5.0,
        'sugarcane': 7.0, 'groundnut': 4.0, 'tomato': 5.5, 'banana': 6.0,
        'default': 5.0
    }
    
    # Critical moisture thresholds by crop
    MOISTURE_THRESHOLD = {
        'rice': 50, 'wheat': 40, 'maize': 35, 'cotton': 35,
        'sugarcane': 45, 'groundnut': 30, 'tomato': 40, 'banana': 50,
        'default': 40
    }
    
    crop_lower = crop.lower()
    daily_etc = CROP_ETC.get(crop_lower, CROP_ETC['default'])
    threshold = MOISTURE_THRESHOLD.get(crop_lower, MOISTURE_THRESHOLD['default'])
    
    # Adjust for temperature
    if temperature > 35:
        daily_etc *= 1.25
        temp_note = "high"
    elif temperature < 20:
        daily_etc *= 0.75
        temp_note = "low"
    else:
        temp_note = "normal"
    
    # Adjust for humidity
    if humidity > 80:
        daily_etc *= 0.85
    elif humidity < 40:
        daily_etc *= 1.15
    
    # Effective rainfall consideration
    effective_rain = min(rainfall_last_7_days * 0.7, daily_etc * 7)
    
    # Calculate deficit
    deficit = max(0, (daily_etc * 7) - effective_rain)
    
    # Volume in mm per hectare (1mm = 10,000 liters/ha)
    total_liters_per_ha = deficit * 10000
    total_liters = total_liters_per_ha * area_hectares
    
    # Urgency assessment
    if soil_moisture < threshold * 0.7:
        urgency = "critical"
        urgency_ta = "அவசரம்"
        action = "Irrigate immediately. Soil moisture critically low."
        action_ta = "உடனனே நீர்ப்பாசனம் செய்யுங்கள். மண் ஈரப்பதம் மிகக் குறைவாக உள்ளது."
    elif soil_moisture < threshold:
        urgency = "needed"
        urgency_ta = "தேவை"
        action = "Irrigation required in the next 1-2 days."
        action_ta = "அடுத்த 1-2 நாட்களில் நீர்ப்பாசனம் தேவை."
    elif soil_moisture < threshold * 1.2:
        urgency = "monitor"
        urgency_ta = "கண்காணிக்கவும்"
        action = "Moisture adequate but monitor daily."
        action_ta = "ஈரப்பதம் போதுமானது, ஆனால் தினமும் கண்காணிக்கவும்."
    else:
        urgency = "adequate"
        urgency_ta = "போதுமானது"
        action = "Soil moisture is good. No irrigation needed today."
        action_ta = "மண் ஈரப்பதம் நல்லது. இன்று நீர்ப்பாசனம் தேவையில்லை."
        total_liters = 0
    
    # Best irrigation times
    best_times = ["6:00 AM - 8:00 AM", "5:00 PM - 7:00 PM"]
    best_times_ta = ["காலை 6:00 - 8:00", "மாலை 5:00 - 7:00"]
    
    # Method recommendation
    if crop_lower in ['banana', 'sugarcane', 'tomato']:
        method = "Drip Irrigation"
        method_ta = "சொட்டு நீர்ப்பாசனம்"
        method_reason = "Saves 40-60% water, prevents leaf diseases"
    elif crop_lower in ['rice']:
        method = "Flood / FMD Irrigation"
        method_ta = "வெள்ள நீர்ப்பாசனம்"
        method_reason = "Traditional flooding with AWD technique saves 30% water"
    else:
        method = "Sprinkler or Furrow"
        method_ta = "தெளிப்பு நீர்ப்பாசனம்"
        method_reason = "Efficient for field crops"
    
    return {
        "urgency": urgency,
        "urgency_ta": urgency_ta,
        "action": action,
        "action_ta": action_ta,
        "water_needed_liters": round(total_liters),
        "water_needed_per_hectare": round(total_liters_per_ha),
        "best_times": best_times,
        "best_times_ta": best_times_ta,
        "irrigation_method": method,
        "irrigation_method_ta": method_ta,
        "method_reason": method_reason,
        "daily_water_need_mm": round(daily_etc, 1),
        "temperature_effect": temp_note,
        "weekly_deficit_mm": round(deficit, 1)
    }

File: irrigation.js
/* ════════════════════════════════
   irrigation.js — Smart Irrigation
   ════════════════════════════════ */

function updateMoistureDisplay(value) {
    document.getElementById('moistureReading').textContent = `${value}%`;
    document.getElementById('moistureFill').style.width = `${value}%`;

    const fill = document.getElementById('moistureFill');
    if (value < 30) {
        fill.style.background = 'linear-gradient(90deg, #ef4444, #f87171)';
    } else if (value < 55) {
        fill.style.background = 'linear-gradient(90deg, #f59e0b, #fbbf24)';
    } else {
        fill.style.background = 'linear-gradient(90deg, var(--green-700), var(--green-400))';
    }
}

document.getElementById('irrigationForm').addEventListener('submit', async (e) => {
    e.preventDefault();
    const btn = e.submitter || e.target.querySelector('button[type="submit"]');
    setLoading(btn, true);

    const payload = {
        crop:           document.getElementById('irrigCrop').value,
        soil_moisture:  parseFloat(document.getElementById('moistureSlider').value),
        temperature:    parseFloat(document.getElementById('irrigTemp').value),
        humidity:       parseFloat(document.getElementById('irrigHumidity').value),
        rainfall_7days: parseFloat(document.getElementById('rainfall7d').value),
        area_hectares:  parseFloat(document.getElementById('farmArea').value)
    };

    try {
        const data = await apiPost('/api/irrigation/advice', payload);
        if (data.success) {
            renderIrrigationResult(data.advice, payload.crop);
            showToast('Irrigation advice ready! / நீர்ப்பாசன ஆலோசனை தயார்!', 'success');
        }
    } catch (err) {
        showToast('Failed to calculate irrigation advice.', 'error');
    } finally {
        setLoading(btn, false);
    }
});

function renderIrrigationResult(advice, crop) {
    const container = document.getElementById('irrigationResults');
    container.classList.remove('hidden');

    const urgencyIcons = {
        critical: '🚨', needed: '⚠️', monitor: '👀', adequate: '✅'
    };

    const actionText = currentLang === 'ta' ? advice.action_ta : advice.action;
    const urgencyText = currentLang === 'ta' ? advice.urgency_ta : advice.urgency.toUpperCase();
    const methodText = currentLang === 'ta' ? advice.irrigation_method_ta : advice.irrigation_method;
    const timeText = currentLang === 'ta' ? advice.best_times_ta : advice.best_times;

    container.innerHTML = `
        <div class="irrigation-result">
            <div class="urgency-banner urgency-${advice.urgency}" style="animation:bounce-in 0.4s ease">
                <span style="font-size:1.5rem">${urgencyIcons[advice.urgency]}</span>
                <div>
                    <div style="font-weight:700">${urgencyText}</div>
                    <div style="font-size:0.85rem;opacity:0.85;font-family:var(--font-tamil)">${actionText}</div>
                </div>
            </div>

            <div class="irrigation-stats">
                <div class="irrig-stat-card">
                    <div class="irrig-stat-value">${advice.water_needed_liters > 0 ? advice.water_needed_liters.toLocaleString() : '—'}</div>
                    <div class="irrig-stat-label">Total Liters Needed</div>
                </div>
                <div class="irrig-stat-card">
                    <div class="irrig-stat-value">${advice.daily_water_need_mm} mm</div>
                    <div class="irrig-stat-label">Daily Water Need</div>
                </div>
                <div class="irrig-stat-card">
                    <div class="irrig-stat-value">${advice.weekly_deficit_mm} mm</div>
                    <div class="irrig-stat-label">Weekly Deficit</div>
                </div>
                <div class="irrig-stat-card">
                    <div class="irrig-stat-value">${advice.temperature_effect === 'high' ? '🌡️ High' : advice.temperature_effect === 'low' ? '❄️ Low' : '✅ Normal'}</div>
                    <div class="irrig-stat-label">Temperature Effect</div>
                </div>
            </div>

            <div style="display:grid;grid-template-columns:1fr 1fr;gap:16px;margin-bottom:20px;">
                <div class="irrig-method-card">
                    <div class="irrig-method-name">💧 ${methodText}</div>
                    <div class="irrig-method-reason">${advice.method_reason}</div>
                </div>
                <div class="irrig-method-card">
                    <div class="irrig-method-name">⏰ Best Time to Water</div>
                    <div style="display:flex;flex-direction:column;gap:4px;margin-top:6px;">
                        ${timeText.map(t => `<span style="font-size:0.85rem;color:var(--green-300);font-family:var(--font-tamil)">${t}</span>`).join('')}
                    </div>
                </div>
            </div>

            <div style="background:rgba(46,171,46,0.06);border-radius:8px;padding:14px;border:1px solid var(--border)">
                <div style="font-size:0.78rem;font-weight:700;text-transform:uppercase;letter-spacing:0.7px;color:var(--text-muted);margin-bottom:8px">
                    💡 Pro Tips for ${crop.charAt(0).toUpperCase() + crop.slice(1)}
                </div>
                <ul style="list-style:none;display:flex;flex-direction:column;gap:6px;">
                    <li style="font-size:0.82rem;color:var(--text-secondary)">• Water at soil level to avoid leaf wetting and fungal risk</li>
                    <li style="font-size:0.82rem;color:var(--text-secondary)">• Avoid irrigating on windy days (>20 km/h) to reduce evaporation loss</li>
                    <li style="font-size:0.82rem;color:var(--text-secondary)">• Check soil moisture with a simple finger test 5cm deep before each irrigation</li>
                </ul>
            </div>
        </div>
    `;
}
